bulldozer/merge.go

```diff
@@ -62,8 +65,49 @@ func (m *GitHubMerger) DeleteHead(ctx context.Context, pullCtx pull.Context) err
 	return errors.WithStack(err)
 }
 
-const MaxPullRequestPollCount = 5
+// PushRestrictionMerger delegates merge operations to different Mergers based
+// on whether or not the pull requests targets a branch with push restrictions.
+type PushRestrictionMerger struct {
+	Normal     Merger
+	Restricted Merger
+}
+
+func NewPushRestrictionMerger(normal, restricted Merger) Merger {
+	return &PushRestrictionMerger{
+		Normal:     normal,
+		Restricted: restricted,
+	}
+}
+
+func (m *PushRestrictionMerger) Merge(ctx context.Context, pullCtx pull.Context, message string, options *github.PullRequestOptions) (string, error) {
+	restricted, err := pullCtx.PushRestrictions(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	if restricted {
+		return m.Restricted.Merge(ctx, pullCtx, message, options)
+	}
+	return m.Normal.Merge(ctx, pullCtx, message, options)
+}
+
+func (m *PushRestrictionMerger) DeleteHead(ctx context.Context, pullCtx pull.Context) error {
+	restricted, err := pullCtx.PushRestrictions(ctx)
+	if err != nil {
+		return err
+	}
+
+	// this is not necessary: the normal client should have delete permissions,
+	// but having the merge user also delete the branch is a better UX
+	if restricted {
+		return m.Restricted.DeleteHead(ctx, pullCtx)
```

Comment:
Do you think it would be valuable here to also log that we're performing a delete action and split it out by adding if we're using the restricted user token or not?

Reply:
The action is logged by the calling code, but I'll add a message about which client was selected.